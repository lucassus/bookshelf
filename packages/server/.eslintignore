coverage
dist
node_modules
jest.config.js
src/**/*.generated.ts
