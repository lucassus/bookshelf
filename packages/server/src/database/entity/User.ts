import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  OneToOne,
  JoinColumn,
  Index,
  CreateDateColumn,
  UpdateDateColumn
} from "typeorm";

import { Avatar } from "./Avatar";

@Entity({ name: "users" })
export class User {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  name: string;

  @Column()
  @Index({ unique: true })
  email: string;

  @Column({ name: "avatar_id" })
  avatarId: number;

  @OneToOne(() => Avatar, { cascade: true, eager: true })
  @JoinColumn({ name: "avatar_id" })
  avatar: Avatar;

  @CreateDateColumn({ name: "created_at" })
  created_at: Date;

  @UpdateDateColumn({ name: "updated_at" })
  updated_at: Date;
}
