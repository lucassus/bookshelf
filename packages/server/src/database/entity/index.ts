export { Author } from "./Author";
export { Avatar } from "./Avatar";
export { Book } from "./Book";
export { BookCopy } from "./BookCopy";
export { User } from "./User";
