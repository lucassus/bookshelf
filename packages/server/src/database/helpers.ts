import { SecureId } from "./SecureId";

type EntityTypes = "Author" | "Book" | "User" | "BookCopy";

export const secureId = new SecureId<EntityTypes>({
  separator: "-"
});
