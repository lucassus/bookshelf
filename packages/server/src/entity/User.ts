import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  OneToOne,
  JoinColumn,
  Index,
  BaseEntity
} from "typeorm";

import { Avatar } from "./Avatar";

// TODO: Add created_at, updated_at

@Entity()
export class User extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  name: string;

  @Column()
  @Index({ unique: true })
  email: string;

  @OneToOne(() => User)
  @JoinColumn({ name: "avatar_id" })
  avatar: Avatar;
}
