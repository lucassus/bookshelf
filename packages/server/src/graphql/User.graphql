type User {
  id: ID!
  name: String!
  info: String
  email: String!
  avatar: Avatar!
  ownedBookCopies: [BookCopy!]!
  borrowedBookCopies: [BookCopy!]!
}

type Avatar {
  image: Image!
  color: String!
}
