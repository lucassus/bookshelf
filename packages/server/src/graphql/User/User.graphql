type User implements Resource & Timestampable {
  id: ExternalID!
  name: String!
  info: String
  email: String!
  avatar: Avatar!
  ownedBookCopies: [BookCopy!]!
  borrowedBookCopies: [BookCopy!]!
  createdAt: String!
  updatedAt: String!
}

type Avatar {
  image: Image!
  color: String!
}

input AvatarInput {
  imagePath: String!
  color: String!
}

input CreateUserInput {
  name: String!
  info: String!
  email: String!
  avatar: AvatarInput!
}

input UpdateUserInput {
  id: ExternalID!
  name: String!
  email: String!
  info: String!
}
