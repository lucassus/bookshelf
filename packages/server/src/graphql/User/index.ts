export { resolvers } from "./User.resolvers";
