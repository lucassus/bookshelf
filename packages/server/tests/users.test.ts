import { ApolloServer, gql } from "apollo-server-express";
import { createTestClient } from "apollo-server-testing";
import { getConnection } from "typeorm";

import { User } from "../src/database/entity/User";
import { secureId } from "../src/database/helpers";
import { createServer } from "../src/server";

let server: ApolloServer;

beforeEach(async () => {
  server = createServer(getConnection());
});

it("fetches users", async () => {
  const { query } = createTestClient(server);

  // When
  const res = await query({
    query: gql`
      query {
        users {
          name
          email
          info
          avatar {
            color
            image {
              url
            }
          }
        }
      }
    `
  });

  // Then
  expect(res.data).not.toBeUndefined();
  expect(res.data!.users).toMatchSnapshot();
});

it("fetches a user", async () => {
  const { query } = createTestClient(server);
  const user = await getConnection().manager.findOneOrFail(User, {
    name: "Bob"
  });

  // When
  const res = await query({
    query: gql`
      query GetUser($id: ID!) {
        user(id: $id) {
          name
          email
          info
          avatar {
            color
            image {
              url
            }
          }
          ownedBookCopies {
            book {
              id
              title
            }
          }
          borrowedBookCopies {
            borrower {
              id
              name
              email
            }
            book {
              id
              title
            }
          }
        }
      }
    `,
    variables: { id: secureId.toExternal(user.id, "User") }
  });

  // Then
  expect(res.data).not.toBeUndefined();
  expect(res.data!.user).toMatchSnapshot();
});
