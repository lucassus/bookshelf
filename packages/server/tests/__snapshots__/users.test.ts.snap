// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches a user 1`] = `
Object {
  "avatar": Object {
    "color": "green",
    "image": Object {
      "url": "http://examples.devmastery.pl/assets/images/avatars/m10.png",
    },
  },
  "borrowedBookCopies": Array [
    Object {
      "book": Object {
        "id": "MTktQm9vaw==",
        "title": "The tower of the swallow",
      },
      "borrower": Object {
        "email": "bob@example.com",
        "id": "Mi1Vc2Vy",
        "name": "Bob",
      },
    },
  ],
  "email": "bob@example.com",
  "info": "Twitter fan. Social media expert. Hardcore explorer. Communicator. Amateur coffee lover.",
  "name": "Bob",
  "ownedBookCopies": Array [
    Object {
      "book": Object {
        "id": "MTYtQm9vaw==",
        "title": "Blood of Elves",
      },
    },
    Object {
      "book": Object {
        "id": "MjAtQm9vaw==",
        "title": "The lady of the lake",
      },
    },
  ],
}
`;

exports[`fetches users 1`] = `
Array [
  Object {
    "avatar": Object {
      "color": "yellow",
      "image": Object {
        "url": "http://examples.devmastery.pl/assets/images/avatars/w13.png",
      },
    },
    "email": "alice@example.com",
    "info": "Food scholar. Incurable tv fanatic. Reader. Typical zombie buff. Gamer. Lifelong creator. Certified organizer.",
    "name": "Alice",
  },
  Object {
    "avatar": Object {
      "color": "green",
      "image": Object {
        "url": "http://examples.devmastery.pl/assets/images/avatars/m10.png",
      },
    },
    "email": "bob@example.com",
    "info": "Twitter fan. Social media expert. Hardcore explorer. Communicator. Amateur coffee lover.",
    "name": "Bob",
  },
  Object {
    "avatar": Object {
      "color": "red",
      "image": Object {
        "url": "http://examples.devmastery.pl/assets/images/avatars/w2.png",
      },
    },
    "email": "celine@example.com",
    "info": "Wannabe pop culture specialist. Tv scholar. Lifelong social media practitioner. Music geek. Falls down a lot.",
    "name": "Celine",
  },
  Object {
    "avatar": Object {
      "color": "blue",
      "image": Object {
        "url": "http://examples.devmastery.pl/assets/images/avatars/m25.png",
      },
    },
    "email": "dan@example.com",
    "info": "Professional problem solver. Devoted explorer. Gamer. Unapologetic bacon guru.",
    "name": "Dan",
  },
]
`;
